daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/World.scala

```diff
@@ -55,4 +58,38 @@ private[validation] class World(packages: PartialFunction[PackageId, Ast.Package
         throw EUnknownDefinition(ctx, LEValue(name))
     }
 
+  def idsInModule(ctx: => Context, pkgId: PackageId, modName: ModuleName): Set[Identifier] =
+    cacheModIdentifiers.getOrElseUpdate(
+      (pkgId, modName),
+      lookupModule(ctx, pkgId, modName).definitions.values.foldLeft(Set.empty[Identifier])(idsInDef)
+    )
+
+  private[this] val cacheModIdentifiers =
+    mutable.Map.empty[(PackageId, ModuleName), Set[Identifier]]
+
+  private[this] def idsInType(acc: Set[Identifier], typ0: Ast.Type): Set[Identifier] =
+    typ0 match {
+      case Ast.TSynApp(typeSynName, _) =>
+        TypeTraversable(typ0).foldLeft(acc + typeSynName)(idsInType)
+      case Ast.TTyCon(typeConName) =>
+        acc + typeConName
+      case otherwise =>
+        TypeTraversable(otherwise).foldLeft(acc)(idsInType)
+    }
+
+  private[this] def idsInExpr(acc: Set[Identifier], expr0: Ast.Expr): Set[Identifier] =
+    expr0 match {
+      case Ast.EVal(valRef) =>
+        acc + valRef
+      case Ast.EAbs(binder @ _, body, ref) =>
```

Comment:
Why would we want to add `ref` as well? I'm convinced that's the right thing to do.